05-mvc-with-rails/rails-sessions-and-auth-with-devise/readme.md

```diff
@@ -0,0 +1,365 @@
+
+Make sure you fork/clone this repo if you would like to follow along. It is not too important to follow along for this portion of this lesson. I encourage you to just watch and use the following code to practice it yourself later.
+
+## Set and Use session key-value pairs
+
+In order to understand sessions, we're going to add some code to our existing artists controller. In `app/controllers/artists_controller.rb`:
+
+```ruby
+# this action will set `test: "this string was set on the testing_session action"` as a key value pair in the session
+def testing_session
+  session[:test] = "this string was set on the testing_session action"
+end
+
+# also add @test = session[:test] to index action
+def index
+  @test = session[:test]
+  @artists = Artist.all
+end
+```
+
+Let's update our `app/views/artists/index.html.erb` to include the use of our new instance variable `@test`. Place the following code somewhere in that file:
+
+```html
+<h1>The value of @test is: <%= @test %></h1>
+```
+
+Now because we created this action, let's update our `config/routes.rb` to listen for a request to that action:
+
+```ruby
+get '/artists/testing_session', to: 'artists#testing_session'
+```
+
+Finally we need to create a view for this action. Lets create that now `$ touch views/artists/testing_session.html.erb`. Let's make this view simple and just put in the following content.
+
+```html
+<h1>A session key value pair is being set</h1>
+```
+
+Let's first navigate to our artists index route at `http://localhost:3000/artists`.
+
+We can see that there is nothing after `The value of @test is: `
+
+Now let's navigate to `http://localhost:3000/artists/testing_session`
+
+Great that's nice, our view was generated, but more importantly we had a key-value pair get set in this request because of the `testing_session` action in our artists controller:
+
+```ruby
+def testing_session
+  session[:test] = "this was set on the testing_session action"
+end
+```
+
+How do we verify this session key-value pair was set? Let's navigate back to artists index path.
+
+Now we can see `The value of @test is: this was set on the testing_session action`
+
+If we open a new incognito window and navigate to `http://localhost:3000/artists`, we'll see the session isn't set in that browser because it's a brand new session.
+
+When we clear our cookies/session in our browser we can see in our index route that `@test` becomes nil again.
+
+> Now that we have the ability to transfer information from request to request. We can preserve state in our application.
+
+## Setting objects in sessions
+Instead of hardcoding strings into our sessions hash, lets see if we can store objects themselves in the session.
+
+Let's update our show action to set a session. In `app/controllers/artists_controller.rb`:
+
+```ruby
+def index
+  @last_viewed_artist = session[:last_viewed_artist]
+  @artists = Artist.all
+end
+
+def show
+  @artist = Artist.find(params[:id])
+  session[:last_viewed_artist] = @artist
```

Comment:
:+1: Storing instances of objects in the session is an anti-pattern that can lead to nasty bugs.